docs/source/trainer/numerics.md

```diff
@@ -0,0 +1,73 @@
+
+For training, the most commonly used low precision formats are IEEE `fp16` and the Brain Floating Point `bf16` format.  While both formats utilize the same number of bits in memory, `bf16` offers a greater numerical range than `fp16` at the expense of being able to represent fewer decimal places, resulting in reduced sensitivity.
+
+Please see the following references for additioanl details on the common number formats:
+
+- `fp32`: [Single-precision Floating-point](https://en.wikipedia.org/wiki/Single-precision_floating-point_format)
+- `fp16`: [Half-precision Floating-point](https://en.wikipedia.org/wiki/Half-precision_floating-point_format)
+- `bf16`: [Brain Floating-point](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format)
+
+### Mixed Precision
+
+Using half-precision number formats can boost model throughput, though not without issue.  The reduced representable range and sensitivity of these formats can cause several problems during training such as:
+
+- Gradient underflow: Very small values can get zeroed out
+- Incorrect weight updates: Small gradient values can lead to no weight updates
+- Activation/Loss overflow: Very large values can roll over to representations of [Infinity](https://en.wikipedia.org/wiki/Infinity#Computing)(`Inf`)or [Not a Number](https://en.wikipedia.org/wiki/NaN) (`NaN`)
+
+The result is reduced model accuracy.
+
+The solution is to perform mixed precision training, where both single and half-precision formats are utilized strategically.  Mixed precision training is performed as follows:
+
+1. Compute the forward pass and loss in half-precision, *except* for computations that can cause activations/loss overflow, which are performed in single-precision (e.g., BatchNorm)
+2. Compute the backpropagation pass in half-precision
+3. To increase numerical sensitivity, convert the model to single-precision and perform the optimizer step, enabling the weight update to be done more precisely
+4. Convert the model back to half-precision
+
+The procedure above mitigates the issues with overflow and imprecise weight updates. 
+
+#### Gradient Scaling
+
+In order to prevent gradient underflow, mixed precision training uses a technique called **Gradient Scaling** where the loss is multiplied by a scaling factor *before* backpropagation is performed (step 2 above). This trick scales the computed gradient values to a range where they cannot underflow.  Once the model has been converted to single-precision following the backward pass, the gradients are divided by the scaling factor to restore the gradient values to the original range before stepping the optimizer.
```

Comment:
Maybe elaborate how gradient scaling automatically tunes itself? It catches nan/inf errors and then adjusts the scaling factor for the next batch. So batches that had nan/infs are skipped.